pep-0484.txt

```diff
@@ -276,6 +276,33 @@ implemented by deferring to ``isinstance(x, collections.abc.Callable)``.
 However, ``isinstance(x, typing.Callable[...])`` is not supported.
 
 
+Decorators
+----------
+
+Decorators can modify the types of the functions or classes they
+decorate. To annotate a decorated declaration with the eventual
+decorated type, put a type comment on the line with the decorator::
+
+  from typing import ContextManager, Generator
+  from contextlib import contextmanager
+
+  class DatabaseSession: ...
+
+  @contextmanager  # type: Callable[[str], ContextManager[DatabaseSession]]
```

Comment:
> Is that the general idea you're getting at?

Yes. Maybe just ``cast`` could be tweaked to be acceptable here. For example:
```python
def cast(*args):
    if len(args) == 2:
        return args[1]
    if len(args) == 1:
        return lambda x: x  # Maybe the type could be stored here on 'x'?
```
Then it could be used like this:
```python
cast(List[int], [])  # Returns []

@cast(Callable[[str], ContextManager[DatabaseSession]])
@contextmanager
def session(...): ...
```
But this is just an idea, maybe it is not very useful.